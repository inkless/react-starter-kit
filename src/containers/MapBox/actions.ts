import { ViewState } from 'react-map-gl';
import { ActionType, createAsyncAction, createStandardAction } from 'typesafe-actions';
import { MapboxState, SimulationGeoFeature } from './types';

/**
 * Three actions for sending request to fetch geocoding
 * request, success, failure
 */
export const fetchOriginGeocodingAction = createAsyncAction(
  '@@MapBox/FETCH_ORIGIN_GEOCODING_REQUEST',
  '@@MapBox/FETCH_ORIGIN_GEOCODING_SUCCESS',
  '@@MapBox/FETCH_ORIGIN_GEOCODING_FAILURE'
)<undefined, MapboxState['origin'], Error>();

/**
 * Action to set map viewport
 */
export const setViewport = createStandardAction('@@MapBox/SET_VIEWPORT')<ViewState>();

/**
 * Action to set deliverSimulationStarted, indicating if simulation has started
 */
export const setDeliverSimulationStarted = createStandardAction(
  '@@MapBox/SET_DELIVER_SIMULATION_STARTED'
)<boolean>();

/**
 * Action to set current deliver simulation's geo features,
 * including routes and points (showing the small car)
 */
export const setDeliverGeoFeatures = createStandardAction('@@MapBox/SET_DELIVER_GEO_FEATURES')<
  SimulationGeoFeature[]
>();

/**
 * Move camera to a courier
 */
export const moveCamera = createStandardAction('@@MapBox/MOVE_CAMERA')<string>();

export type MapboxAction =
  | ActionType<typeof fetchOriginGeocodingAction>
  | ActionType<typeof setViewport>
  | ActionType<typeof setDeliverSimulationStarted>
  | ActionType<typeof setDeliverGeoFeatures>
  | ActionType<typeof moveCamera>;
