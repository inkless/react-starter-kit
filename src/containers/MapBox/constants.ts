export const MAPBOX_API_URL = 'https://api.mapbox.com';
export const DEFAULT_ORIGIN = {
  name: 'Home',
  address: '1800 Marine Street, Santa Monica, CA 90405',
};

/**
 * source for mapbox, all our layers is sharing one source,
 * so it has better performance
 */
export const MAPBOX_SOURCE_NAME = 'react-starter-kit';

/**
 * Simulation animation interval
 * Technically using requestAnimationFrame can make animation smoothier,
 * but that consumes too much CPU/GPU, using an interval instead
 */
export const ANIMATION_INTERVAL = 500;
