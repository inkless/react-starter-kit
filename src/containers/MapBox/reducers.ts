import { getType } from 'typesafe-actions';

import { MapboxState } from './types';
import {
  MapboxAction,
  fetchOriginGeocodingAction,
  setViewport,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
  moveCamera,
} from './actions';
import { Feature, Point } from '@turf/helpers';

const defaultState: MapboxState = {
  origin: {
    address: '',
    coordinates: [0, 0],
  },
  viewport: {
    longitude: 0,
    latitude: 0,
    zoom: 11.5,
  },
  deliverSimulationStarted: false,
  deliverSimulationGeoFeatures: [],
};

export function mapboxReducer(state = defaultState, action: MapboxAction): MapboxState {
  switch (action.type) {
    case getType(fetchOriginGeocodingAction.success): {
      const [longitude, latitude] = action.payload.coordinates;
      return {
        ...state,
        origin: action.payload,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    case getType(setViewport): {
      // Note: this is needed, ReactMapGL will send width and height with the
      // viewport, in order to allow map resize properly when window resize,
      // we need to remove this width, height from payload
      const { width, height, ...viewport } = action.payload;
      return {
        ...state,
        viewport,
      };
    }
    case getType(setDeliverSimulationStarted):
      return {
        ...state,
        deliverSimulationStarted: action.payload,
      };
    case getType(setDeliverGeoFeatures):
      return {
        ...state,
        deliverSimulationGeoFeatures: action.payload,
      };
    case getType(moveCamera): {
      const courierPoint = state.deliverSimulationGeoFeatures.find(
        feature => feature.properties.id === action.payload
      ) as Feature<Point, any> | undefined;

      if (!courierPoint || !courierPoint.geometry) {
        return state;
      }
      const [longitude, latitude] = courierPoint.geometry.coordinates;
      return {
        ...state,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    default:
      return state;
  }
}
