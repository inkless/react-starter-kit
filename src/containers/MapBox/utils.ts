import request from '@src/utils/request';
import { lineString, Feature } from '@turf/helpers';
import turfLength from '@turf/length';
import { decode } from '@mapbox/polyline';

import { StringAnyMap } from '@src/types/utils';
import { MAPBOX_API_URL } from './constants';

const ACCESS_TOKEN = process.env.MapboxAccessToken;
const POLYLINE_PRECISION = 5;
const SECONDS_IN_HOUR = 3600;
const SPEED_BASE = 20;

/**
 * Fetch driving route from origin to destination
 *
 * @param origin
 * @param destination
 */
export async function fetchDrivingRoute(origin: [number, number], destination: [number, number]) {
  const baseUrl = `${MAPBOX_API_URL}/directions/v5/mapbox/driving-traffic`;
  const coordinates = `${origin.join(',')};${destination.join(',')}`;
  const url = `${baseUrl}/${coordinates}?steps=true&overview=full&access_token=${ACCESS_TOKEN}`;
  const data = await request(url);
  return data.routes[0];
}

/**
 * Given any address, find the geocoding
 * @param address
 */
export async function fetchGeocoding(address: string) {
  const url = `${MAPBOX_API_URL}/geocoding/v5/mapbox.places/${encodeURIComponent(
    address
  )}.json?access_token=${ACCESS_TOKEN}`;
  const data = await request(url);
  return data.features[0].center;
}

/**
 * Generate a speed between 20 to 25 miles per hour
 * @returns miles per second
 */
export function generateRandomSpeed() {
  return (Math.random() * 5 + SPEED_BASE) / SECONDS_IN_HOUR;
}

/**
 * Decode geometry and return geojson route
 * @param geometry
 * @returns geojson lineString
 */
export function createRouteFromGeometry(geometry: string, options?: StringAnyMap) {
  const decoded = decode(geometry, POLYLINE_PRECISION).map(c => {
    return c.reverse();
  });

  return lineString(decoded, options);
}

/**
 * Given a route, calculate the distance of it
 * @param route
 * @returns distance in miles
 */
export function calculateRouteLength(route: Feature) {
  return turfLength(route, { units: 'miles' });
}
