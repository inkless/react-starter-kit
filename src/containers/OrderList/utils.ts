import { groupBy } from '@src/utils/array';
import { PendingOrder } from './types';

/**
 * reduce all orders' dispatchTime by 1 second,
 * move them to delivering if dispatchTime is 0
 *
 * @param orders - pending orders
 * @returns
 */
export function tickOneSecond(orders: PendingOrder[]) {
  const updatedOrders = orders.map(order => ({
    ...order,
    dispatchTime: order.dispatchTime - 1,
  }));

  const { true: delivering = [], false: pending = [] } = groupBy(
    updatedOrders,
    ({ dispatchTime }) => dispatchTime === 0
  );

  return {
    delivering,
    pending,
  };
}
