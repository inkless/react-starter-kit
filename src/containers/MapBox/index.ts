import { connect } from 'react-redux';

import MapBox from '@src/components/MapBox';
import { RootState } from '@src/store';
import { fetchOriginOperation } from './operations';
import { setViewport } from './actions';
import { mapStyleSelector, getDeliveringOrders } from './selectors';

const mapStateToProps = (state: RootState) => ({
  ...state.mapbox,
  mapStyle: mapStyleSelector(state),
  deliveringOrders: getDeliveringOrders(state),
});

const mapDispatchToProps = {
  fetchOriginOperation,
  setViewport,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MapBox);
