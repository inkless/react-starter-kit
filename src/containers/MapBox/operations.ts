import lineSliceAlong from '@turf/line-slice-along';
import { point } from '@turf/helpers';

import { ThunkResult } from '@src/store';
import { DeliveringOrder, PendingOrder } from '@src/containers/OrderList/types';
import {
  startDeliveringOrder,
  fetchDestinationAction,
  orderDelivered,
} from '@src/containers/OrderList/actions';
import {
  fetchOriginGeocodingAction,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
} from './actions';
import { DEFAULT_ORIGIN, ANIMATION_INTERVAL } from './constants';
import {
  fetchGeocoding,
  fetchDrivingRoute,
  generateRandomSpeed,
  createRouteFromGeometry,
  calculateRouteLength,
} from './utils';
import { SimulationGeoFeature } from '@src/containers/MapBox/types';

/**
 * @param address A query string of address
 */
export function fetchOriginOperation(address = DEFAULT_ORIGIN.address): ThunkResult<void> {
  return async function(dispatch) {
    const { request, success, failure } = fetchOriginGeocodingAction;
    dispatch(request());
    try {
      const coordinates = await fetchGeocoding(address);
      dispatch(
        success({
          address,
          coordinates,
        })
      );
    } catch (e) {
      dispatch(failure(e));
    }
  };
}

/**
 * Send API request to fetch pendingOrder's geo infomation,
 * then send an action to start delivering order.
 * If simulation is not started, start simulation.
 */
export function fetchDestinationOperation({
  dispatchTime,
  ...order
}: PendingOrder): ThunkResult<void> {
  return async function(dispatch, getState) {
    const {
      mapbox: { origin, deliverSimulationStarted },
    } = getState();
    dispatch(fetchDestinationAction.request());
    try {
      // fetch coordinates first, then fetch driving route info
      const coordinates = await fetchGeocoding(order.destination);
      const { geometry } = await fetchDrivingRoute(origin.coordinates, coordinates);
      const route = createRouteFromGeometry(geometry, {
        id: `route-${order.id}`,
        route: 'destination',
      });

      const deliveringOrder: DeliveringOrder = {
        ...order,
        coordinates,
        route,
        startTime: new Date(),
        speed: generateRandomSpeed(),
        distance: calculateRouteLength(route),
      };

      // start simulation if not started
      if (!deliverSimulationStarted) {
        dispatch(startSimulatingDeliver());
      }

      dispatch(startDeliveringOrder(deliveringOrder));
    } catch (e) {
      dispatch(fetchDestinationAction.failure(e));
    }
  };
}

/**
 * startSimulatingDeliver
 */
export function startSimulatingDeliver(): ThunkResult<void> {
  return function(dispatch, getState) {
    // do nothing if it's already started
    if (getState().mapbox.deliverSimulationStarted) {
      return;
    }

    dispatch(setDeliverSimulationStarted(true));

    /**
     * Notice: To have a even smoothier animation,
     * we could use requestAnimationFrame.
     * However, that consume much more CPU/GPU, I'll keep it simple here.
     */
    const interval = setInterval(() => {
      const {
        orderList: { delivering, pending },
      } = getState();

      const features: SimulationGeoFeature[] = [];
      delivering.forEach(order => {
        const { id, name, speed, startTime, route, distance } = order;

        const traveledDistance = speed * ((Date.now() - startTime.valueOf()) / 1000);

        // Use traveledDistance, calculate current route
        const currentRoute = lineSliceAlong(route, 0, traveledDistance, { units: 'miles' });
        currentRoute.properties = {
          ...currentRoute.properties,
          id: `current-route-${id}`,
          route: 'current',
        };

        if (currentRoute.geometry !== null) {
          // the position of courier should be the last point of currentRoute
          const lastCoordinate = currentRoute.geometry.coordinates.slice(-1)[0];
          const currentPoint = point(lastCoordinate, {
            name,
            id: `current-point-${id}`,
          });

          // push point/route to features
          features.push(currentPoint, currentRoute);
        }

        // dispatch orderDelivered when travel distance is greater than
        // destination distance
        if (traveledDistance >= distance) {
          dispatch(
            orderDelivered({
              ...order,
              endTime: new Date(),
            })
          );
        }
      });

      // dispatch geo features
      if (features.length) {
        dispatch(setDeliverGeoFeatures(features));
      }

      // when there's no more pending, and delivering,
      // stop simulation
      if (!pending.length && !delivering.length) {
        clearInterval(interval);
      }
    }, ANIMATION_INTERVAL);
  };
}
