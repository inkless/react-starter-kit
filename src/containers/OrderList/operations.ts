import { ThunkResult } from '@src/store';
import initialOrders from '@src/data/orders.json';
import { updatePendingOrders, setSimulatingOrdersStarted } from './actions';
import { tickOneSecond } from './utils';
import { fetchDestinationOperation } from '@src/containers/MapBox/operations';

/**
 * Start to update pending orders, subtract dispatchTime by 1 per one second
 */
export function startSimulatingOrders(): ThunkResult<void> {
  return function(dispatch, getState) {
    const { orderList } = getState();

    if (orderList.simulatingOrdersStarted) {
      return;
    }

    // when orderList does not have any data, we fetch it from initialOrders
    if (!orderList.pending.length && !orderList.delivering.length && !orderList.delivered.length) {
      initialOrders.sort((a, b) => a.dispatchTime - b.dispatchTime);
      dispatch(updatePendingOrders(initialOrders));
    }

    const interval = setInterval(() => {
      const { orderList } = getState();
      if (!orderList.pending.length) {
        clearInterval(interval);
        return;
      }

      // tick one second, update delivering and pending orders
      const { pending, delivering } = tickOneSecond(orderList.pending);
      dispatch(updatePendingOrders(pending));

      // if there is delivering order after one second tick
      // start an operation to fetch destination info
      if (delivering && delivering.length) {
        delivering.forEach(order => {
          dispatch(fetchDestinationOperation(order));
        });
      }
    }, 1000);

    dispatch(setSimulatingOrdersStarted(true));
  };
}
