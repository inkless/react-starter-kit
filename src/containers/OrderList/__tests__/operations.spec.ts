import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { fetchDestinationOperation } from '@src/containers/MapBox/operations';
import { createInitialState } from '@src/utils/test-helper';
import { pending } from '@src/components/OrderList/__tests__/__fixture__/orders';
import { startSimulatingOrders } from '../operations';

const rootState = createInitialState();
const mockStore = configureStore([thunk]);

jest.useFakeTimers();
jest.mock('@src/containers/MapBox/operations');

describe('OrderList Operations', () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  describe('#fetchOriginOperation', () => {
    it('should dispatch correctly when no pending', () => {
      const store = mockStore(rootState);
      store.dispatch<any>(startSimulatingOrders());
      expect(store.getActions()).toMatchSnapshot();

      // tick one second
      jest.runTimersToTime(1000);
      expect(clearInterval).toHaveBeenCalledTimes(1);
    });

    it('should dispatch correctly when pending, delivering exist', () => {
      const store = mockStore({
        ...rootState,
        orderList: {
          ...rootState.orderList,
          pending,
        },
      });
      store.dispatch<any>(startSimulatingOrders());
      jest.runTimersToTime(1000);

      expect(store.getActions()).toMatchSnapshot();
    });

    it('should dispatch correctly when there is new delivering', () => {
      const updatedPending = [...pending];
      updatedPending[0] = {
        ...updatedPending[0],
        dispatchTime: 1,
      };
      const store = mockStore({
        ...rootState,
        orderList: {
          ...rootState.orderList,
          pending: updatedPending,
        },
      });

      (fetchDestinationOperation as jest.Mock<any>).mockReturnValue({
        type: 'mock',
      });
      store.dispatch<any>(startSimulatingOrders());
      jest.runTimersToTime(1000);

      expect(fetchDestinationOperation).toHaveBeenCalledWith({
        ...updatedPending[0],
        dispatchTime: 0,
      });
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
