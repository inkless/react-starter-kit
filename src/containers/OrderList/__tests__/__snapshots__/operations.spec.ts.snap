// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList Operations #fetchOriginOperation should dispatch correctly when no pending 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Array [
      Object {
        "destination": "1416 4th St, Santa Monica, CA 90401",
        "dispatchTime": 3,
        "id": "5sSNvC",
        "name": "Cottage Cheese",
      },
      Object {
        "destination": "1630 Ocean Park Blvd, Santa Monica, CA 90405",
        "dispatchTime": 13,
        "id": "4WVKZu",
        "name": "Cobb Salad",
      },
      Object {
        "destination": "2208 Sawtelle Blvd, Los Angeles, CA 90064",
        "dispatchTime": 16,
        "id": "tG7Mtb",
        "name": "Beef Stew",
      },
      Object {
        "destination": "400 World Way, Los Angeles, CA 90045",
        "dispatchTime": 24,
        "id": "ZEUn6r",
        "name": "Chunky Monkey",
      },
      Object {
        "destination": "10117 Washington Blvd, Culver City, CA 90232",
        "dispatchTime": 24,
        "id": "z57hNM",
        "name": "Beef Hash",
      },
      Object {
        "destination": "11777 San Vicente Blvd., Los Angeles, CA 90049",
        "dispatchTime": 26,
        "id": "3hWiEm",
        "name": "MeatLoaf",
      },
      Object {
        "destination": "4720 Admiralty Way, Marina Del Rey, CA 90292",
        "dispatchTime": 35,
        "id": "uoierx",
        "name": "Acai Bowl",
      },
      Object {
        "destination": "1323 Montana Ave, Santa Monica, CA 90403",
        "dispatchTime": 38,
        "id": "RoYGTg",
        "name": "Cookie Dough",
      },
      Object {
        "destination": "267 N Canon Dr, Beverly Hills, CA 90210",
        "dispatchTime": 39,
        "id": "UWkU5v",
        "name": "Mint Chocolate Ice Cream",
      },
      Object {
        "destination": "3306 Lincoln Blvd, Santa Monica, CA 90405",
        "dispatchTime": 40,
        "id": "2ndwz7",
        "name": "Vegan Pizza",
      },
      Object {
        "destination": "2911 Main St, Santa Monica, CA 90405",
        "dispatchTime": 41,
        "id": "VrXRDR",
        "name": "Arugula",
      },
      Object {
        "destination": "193 Culver Blvd, Playa Del Rey, CA 90293",
        "dispatchTime": 47,
        "id": "7Ur6qT",
        "name": "Chicken",
      },
      Object {
        "destination": "2055 Sawtelle Blvd, Los Angeles, CA 90025",
        "dispatchTime": 49,
        "id": "WFG4B3",
        "name": "Snow Cone",
      },
      Object {
        "destination": "1321 3rd St Promenade, Santa Monica, CA 90401",
        "dispatchTime": 54,
        "id": "tr5K86",
        "name": "Fudge Ice Cream Cake",
      },
      Object {
        "destination": "1303 Westwood Blvd, Los Angeles, CA 90024",
        "dispatchTime": 63,
        "id": "5kyaoG",
        "name": "Cheese",
      },
      Object {
        "destination": "13455 Maxella Ave, Ste 250, Marina Del Rey, CA 90292",
        "dispatchTime": 66,
        "id": "n5byZt",
        "name": "Fresh Fruit",
      },
      Object {
        "destination": "2323 Wilshire Blvd, Santa Monica, CA 90403",
        "dispatchTime": 73,
        "id": "aS4EFD",
        "name": "Chocolate Gelato",
      },
      Object {
        "destination": "1522 Abbott Kinney Blvd, Los Angeles, CA 90291",
        "dispatchTime": 78,
        "id": "2YVKCU",
        "name": "Spinach Omelet",
      },
      Object {
        "destination": "1716 S Sepulveda Blvd, Los Angeles, CA 90025",
        "dispatchTime": 93,
        "id": "7FNaQi",
        "name": "Pork Chop",
      },
      Object {
        "destination": "2906 Main St, Santa Monica, CA 90405",
        "dispatchTime": 94,
        "id": "woqBZv",
        "name": "Pad See Ew",
      },
      Object {
        "destination": "13455 Maxella Ave, Marina Del Rey, CA 90292",
        "dispatchTime": 98,
        "id": "EUtSqE",
        "name": "McFlury",
      },
      Object {
        "destination": "220 Rose Ave, Venice, CA 90291",
        "dispatchTime": 105,
        "id": "oP6QHd",
        "name": "Yogurt",
      },
      Object {
        "destination": "101 Santa Monica Blvd, Santa Monica, CA 90401",
        "dispatchTime": 109,
        "id": "8yJnGe",
        "name": "Cranberry Salad",
      },
      Object {
        "destination": "210 Wilshire Blvd, Santa Monica, Santa Monica, CA 90401",
        "dispatchTime": 109,
        "id": "sKar6C",
        "name": "Pickles",
      },
      Object {
        "destination": "9409 Venice Blvd, Culver City, CA 90232",
        "dispatchTime": 110,
        "id": "EJyQaE",
        "name": "Kale Salad",
      },
      Object {
        "destination": "2310 Pico Blvd, Santa Monica, CA 90405",
        "dispatchTime": 112,
        "id": "PQnWkG",
        "name": "Milk",
      },
      Object {
        "destination": "3918 Van Buren Pl, Culver City, CA 90232",
        "dispatchTime": 112,
        "id": "7VD4xy",
        "name": "Orange Chicken",
      },
      Object {
        "destination": "8115 W Manchester Ave, Playa Del Rey, CA 90293",
        "dispatchTime": 119,
        "id": "Fv8AdV",
        "name": "Coke",
      },
      Object {
        "destination": "185 Culver Blvd, Playa Del Rey, CA 90293",
        "dispatchTime": 119,
        "id": "9H4whD",
        "name": "Banana Split",
      },
      Object {
        "destination": "10113 Washington Blvd, Culver City, CA 90232",
        "dispatchTime": 120,
        "id": "AinKVi",
        "name": "Pastrami Sandwich",
      },
    ],
    "type": "@@OrderList/UPDATE_PENDING_ORDERS",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "@@OrderList/SIMULATING_ORDERS_STARTED",
  },
]
`;

exports[`OrderList Operations #fetchOriginOperation should dispatch correctly when pending, delivering exist 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": true,
    "type": "@@OrderList/SIMULATING_ORDERS_STARTED",
  },
  Object {
    "meta": undefined,
    "payload": Array [
      Object {
        "destination": "185 Culver Blvd, Playa Del Rey, CA 90293",
        "dispatchTime": 118,
        "id": "9H4whD",
        "name": "Banana Split",
      },
      Object {
        "destination": "13455 Maxella Ave, Marina Del Rey, CA 90292",
        "dispatchTime": 97,
        "id": "EUtSqE",
        "name": "McFlury",
      },
      Object {
        "destination": "4720 Admiralty Way, Marina Del Rey, CA 90292",
        "dispatchTime": 34,
        "id": "uoierx",
        "name": "Acai Bowl",
      },
      Object {
        "destination": "220 Rose Ave, Venice, CA 90291",
        "dispatchTime": 104,
        "id": "oP6QHd",
        "name": "Yogurt",
      },
      Object {
        "destination": "2323 Wilshire Blvd, Santa Monica, CA 90403",
        "dispatchTime": 72,
        "id": "aS4EFD",
        "name": "Chocolate Gelato",
      },
      Object {
        "destination": "1630 Ocean Park Blvd, Santa Monica, CA 90405",
        "dispatchTime": 12,
        "id": "4WVKZu",
        "name": "Cobb Salad",
      },
      Object {
        "destination": "1416 4th St, Santa Monica, CA 90401",
        "dispatchTime": 2,
        "id": "5sSNvC",
        "name": "Cottage Cheese",
      },
      Object {
        "destination": "8115 W Manchester Ave, Playa Del Rey, CA 90293",
        "dispatchTime": 118,
        "id": "Fv8AdV",
        "name": "Coke",
      },
      Object {
        "destination": "2055 Sawtelle Blvd, Los Angeles, CA 90025",
        "dispatchTime": 48,
        "id": "WFG4B3",
        "name": "Snow Cone",
      },
      Object {
        "destination": "2906 Main St, Santa Monica, CA 90405",
        "dispatchTime": 93,
        "id": "woqBZv",
        "name": "Pad See Ew",
      },
    ],
    "type": "@@OrderList/UPDATE_PENDING_ORDERS",
  },
]
`;

exports[`OrderList Operations #fetchOriginOperation should dispatch correctly when there is new delivering 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": true,
    "type": "@@OrderList/SIMULATING_ORDERS_STARTED",
  },
  Object {
    "meta": undefined,
    "payload": Array [
      Object {
        "destination": "13455 Maxella Ave, Marina Del Rey, CA 90292",
        "dispatchTime": 97,
        "id": "EUtSqE",
        "name": "McFlury",
      },
      Object {
        "destination": "4720 Admiralty Way, Marina Del Rey, CA 90292",
        "dispatchTime": 34,
        "id": "uoierx",
        "name": "Acai Bowl",
      },
      Object {
        "destination": "220 Rose Ave, Venice, CA 90291",
        "dispatchTime": 104,
        "id": "oP6QHd",
        "name": "Yogurt",
      },
      Object {
        "destination": "2323 Wilshire Blvd, Santa Monica, CA 90403",
        "dispatchTime": 72,
        "id": "aS4EFD",
        "name": "Chocolate Gelato",
      },
      Object {
        "destination": "1630 Ocean Park Blvd, Santa Monica, CA 90405",
        "dispatchTime": 12,
        "id": "4WVKZu",
        "name": "Cobb Salad",
      },
      Object {
        "destination": "1416 4th St, Santa Monica, CA 90401",
        "dispatchTime": 2,
        "id": "5sSNvC",
        "name": "Cottage Cheese",
      },
      Object {
        "destination": "8115 W Manchester Ave, Playa Del Rey, CA 90293",
        "dispatchTime": 118,
        "id": "Fv8AdV",
        "name": "Coke",
      },
      Object {
        "destination": "2055 Sawtelle Blvd, Los Angeles, CA 90025",
        "dispatchTime": 48,
        "id": "WFG4B3",
        "name": "Snow Cone",
      },
      Object {
        "destination": "2906 Main St, Santa Monica, CA 90405",
        "dispatchTime": 93,
        "id": "woqBZv",
        "name": "Pad See Ew",
      },
    ],
    "type": "@@OrderList/UPDATE_PENDING_ORDERS",
  },
  Object {
    "type": "mock",
  },
]
`;
